include/hipSYCL/sycl/libkernel/half.hpp

```diff
@@ -149,72 +152,128 @@ class half {
+        return __acpp_sscp_half_lt(a._data, b._data),
+        return __hlt(fp16::as_cuda_half(a._data), fp16::as_cuda_half(b._data)),
+        return fp16::builtin_less_than(a._data, b._data))
+    }
+  
+    ACPP_UNIVERSAL_TARGET
+    friend bool operator<=(const half& a, const half& b) noexcept {
+      __acpp_backend_switch(
+        return fp16::builtin_less_than_equal(a._data, b._data),
+        return __acpp_sscp_half_lte(a._data, b._data),
+        return __hle(fp16::as_cuda_half(a._data), fp16::as_cuda_half(b._data)),
+        return fp16::builtin_less_than_equal(a._data, b._data))
+    }
+  
+    ACPP_UNIVERSAL_TARGET
+    friend bool operator>(const half& a, const half& b) noexcept {
+      __acpp_backend_switch(
+        return fp16::builtin_greater_than(a._data, b._data),
+        return __acpp_sscp_half_gt(a._data, b._data),
+        return __hgt(fp16::as_cuda_half(a._data), fp16::as_cuda_half(b._data)),
+        return fp16::builtin_greater_than(a._data, b._data))
+    }
+  
+    ACPP_UNIVERSAL_TARGET
+    friend bool operator>=(const half& a, const half& b) noexcept {
+      __acpp_backend_switch(
+        return fp16::builtin_greater_than_equal(a._data, b._data),
+        return __acpp_sscp_half_gte(a._data, b._data),
+        return __hge(fp16::as_cuda_half(a._data), fp16::as_cuda_half(b._data)),
+        return fp16::builtin_greater_than_equal(a._data, b._data))
+    }
+  
+#define OP_FOR_TYPE(op, type)                                         \
+  friend bool operator op(const half& a, const type& b) {             \
+    return a op half(b);                                              \
+  }                                                                   \
+                                                                      \
+  friend bool operator op(const type& a, const half& b) {             \
+    return half(a) op b;                                              \
   }
+  
+#define OP(op)                                                        \
+  OP_FOR_TYPE(op, int)                                                \
+  OP_FOR_TYPE(op, unsigned int)                                       \
+  OP_FOR_TYPE(op, long)                                               \
+  OP_FOR_TYPE(op, long long)                                          \
+  OP_FOR_TYPE(op, unsigned long)                                      \
+  OP_FOR_TYPE(op, unsigned long long)                                 \
+  OP_FOR_TYPE(op, float)                                              \
+  OP_FOR_TYPE(op, double)
 
-  ACPP_UNIVERSAL_TARGET
-  friend bool operator<=(const half& a, const half& b) noexcept {
-    __acpp_backend_switch(
-      return fp16::builtin_less_than_equal(a._data, b._data),
-      return __acpp_sscp_half_lte(a._data, b._data),
-      return __hle(fp16::as_cuda_half(a._data), fp16::as_cuda_half(b._data)),
-      return fp16::builtin_less_than_equal(a._data, b._data))
-  }
+  OP(<)
+  OP(<=)
+  OP(>)
+  OP(>=)
+  OP(==)
+  OP(!=)
```

Comment:
Is it actually correct to implement these operators by casting everything to `half`, even when this causes loss of information? For example, when comparing with `double`, I would expect the `half` to be cast to `double` and the comparison being carried out with double precision.

Reply:
I decided to rip out all the half changes from this PR and put them in #1710.